refactor(forms): added control name in console errors

fdonzello opened the pull request:
## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
Right now, when a form control is used without a parent form group, we get an error not reporting the form control name affected by the issue.
With long forms, it could take a lot of time to get the affected form control.
<img width="905" alt="Screenshot 2024-04-18 at 12 19 31" src="https://github.com/angular/angular/assets/763474/59eb4868-5f5a-4f0f-bcbf-363d3a3ca5f8">


## What is the new behavior?
The console error has been improved by reporting either the form control name or index affected by the issue.
<img width="902" alt="Screenshot 2024-04-18 at 16 06 44" src="https://github.com/angular/angular/assets/763474/9f9c0ee8-ca3b-4eba-97e0-79e96a11d980">



## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


## Other information
This is a starting point and can be further improved or scaled to other methods involved in form wrong usage.

JeanMeche commented on packages/forms/src/directives/reactive_directives/form_control_name.ts:
```diff
@@ -6,15 +6,15 @@
  */
 
-import {Directive, EventEmitter, forwardRef, Host, Inject, Input, OnChanges, OnDestroy, Optional, Output, Provider, Self, SimpleChanges, SkipSelf, ɵWritable as Writable} from '@angular/core';
+import {Directive, EventEmitter, forwardRef, Host, Inject, Input, OnChanges, OnDestroy, Optional, Output, Provider, Self, SimpleChanges, SkipSelf, ɵWritable as Writable,} from '@angular/core';
```
can you revert all the non necessary formating changes, so that the diff only contains your changes ?

AndrewKushnir commented on packages/forms/src/directives/reactive_errors.ts:
```diff
@@ -10,18 +10,20 @@ import {ɵRuntimeError as RuntimeError} from '@angular/core';
 import {RuntimeErrorCode} from '../errors';
 
-import {formArrayNameExample, formControlNameExample, formGroupNameExample, ngModelGroupExample} from './error_examples';
+import {formArrayNameExample, formControlNameExample, formGroupNameExample, ngModelGroupExample,} from './error_examples';
```
Are those changes auto-generated by the linter?
```suggestion
import {formArrayNameExample, formControlNameExample, formGroupNameExample, ngModelGroupExample} from './error_examples';
```

fdonzello replied:
yep, the one running on pre (or post?) hook

AndrewKushnir commented on packages/forms/src/directives/reactive_errors.ts:
```diff
@@ -10,18 +10,20 @@ import {ɵRuntimeError as RuntimeError} from '@angular/core';
       `formControlName must be used with a parent formGroup directive.  You'll want to add a formGroup
       directive and pass it an existing FormGroup instance (you can create one in your class).
 
+      Affected Form Control (name or index): ${controlName ?? 'unavailable'}
```
I'd propose checking the presence of the control name and don't include any extra text if it's not available.

fdonzello replied:
done

AndrewKushnir commented:
@fdonzello could you please also update the commit message (via `git commit --amend`) to make it start with `refactor(forms): `?

fdonzello committed:
refactor(forms): added control name in console errors

When a formControlName is used without a parent formGroup, an error is
logged in the console. Before this commit, there was no information
about which control had the issue. Now, it's reported and the
troubleshoot is much faster.

fdonzello commented:
> @fdonzello could you please also update the commit message (via `git commit --amend`) to make it start with `refactor(forms): `?

done :)

AndrewKushnir commented on packages/forms/src/directives/reactive_errors.ts:
```diff
@@ -13,15 +13,26 @@ import {RuntimeErrorCode} from '../errors';
+}
+
+function describeFormControl(controlName: string|number|null): string {
+  if (controlName) {
+    return `Affected Form Control (name or index): "${controlName}"`;
+  }
```
We should update the condition to make sure that we display extra info for `0` as a value. Can we also infer whether it's an index or a name based on the value type? 
```suggestion
  if (controlName != null /* checks both `null` and `undefined` */) {
    return `Affected Form Control ${typeof controlName === 'string' ? 'name' : 'index'}): "${controlName}"`;
  }
```

fdonzello replied:
done, hopefully ok

AndrewKushnir commented on packages/forms/src/directives/reactive_errors.ts:
```diff
@@ -13,15 +13,26 @@ import {RuntimeErrorCode} from '../errors';
+      `formControlName must be used with a parent formGroup directive. You'll want to add a formGroup
       directive and pass it an existing FormGroup instance (you can create one in your class).
 
+      ${describeFormControl(controlName)}
```
Thanks for the update. I'd be also great to add some tests (or expand existing tests) to make sure we have a correct formatting (new lines) when this extra info is present and when it's not present.

fdonzello replied:
Added tests covering:

- **null** form control name
- **index** form control name
- **string** form control name

In order to extend existing tests, I had to change the `toThrowError` part to be able to test the message for multiple outputs. Not sure if it's okay:

```typescript

expect(() => fixture.detectChanges()).toThrowMatching((e: Error) => {
           if (!e.message.includes(
                   `formControlName must be used with a parent formGroup directive`)) {
             return false;
           }

           if (!e.message.includes(`Affected Form Control index: "0"`)) {
             return false;
           }

           return true;
         });

```

AndrewKushnir commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -3881,11 +3881,65 @@ describe('reactive forms integration tests', () => {
+    it('should throw, without indicating the affected form control, if formControlName is used without a control container', () => {
+      TestBed.overrideComponent(FormGroupComp, {
+        set: {
+          template: `
+          <input type="text" formControlName>
+        `,
```
```suggestion
          template: `
            <input type="text" formControlName>
          `,
```

AndrewKushnir commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -3881,11 +3881,65 @@ describe('reactive forms integration tests', () => {
+    it('should throw if formControlName, with an index, is used without a control container', () => {
+      TestBed.overrideComponent(FormGroupComp, {
+        set: {
+          template: `
+          <input type="text" [formControlName]="0">
+        `,
```
```suggestion
          template: `
            <input type="text" [formControlName]="0">
          `,
```

AndrewKushnir reviewed (commented):
@fdonzello looks great, thanks for addressing the feedback. I've just left a couple minor comments related to formatting in tests.

fdonzello commented:
> @fdonzello looks great, thanks for addressing the feedback. I've just left a couple minor comments related to formatting in tests.

@AndrewKushnir strange that `yarn ng-dev format files packages/forms/test/reactive_integration_spec.ts` doesn't apply the formatting :(

AndrewKushnir commented:
@fdonzello thanks for addressing the feedback! 👍 This PR is now in the merge queue.

AndrewKushnir commented:
This PR was merged into the repository by commit c001b05c771e5e59e94320e21f23030a54ee274b.
